sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/storage/GcsResourceId.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.storage;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import org.apache.beam.sdk.io.fs.ResolveOptions;
+import org.apache.beam.sdk.io.fs.ResolveOptions.StandardResolveOptions;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.beam.sdk.util.gcsfs.GcsPath;
+
+/**
+ * {@link ResourceId} implementation for Google Cloud Storage.
+ */
+public class GcsResourceId implements ResourceId {
+
+  private final GcsPath gcsPath;
+
+  static GcsResourceId fromGcsPath(GcsPath gcsPath) {
+    checkNotNull(gcsPath, "gcsPath");
+    return new GcsResourceId(gcsPath);
+  }
+
+  private GcsResourceId(GcsPath gcsPath) {
+    this.gcsPath = gcsPath;
+  }
+
+  @Override
+  public ResourceId resolve(String other, ResolveOptions resolveOptions) {
```

Comment:
Added more details in javadoc.

We already have GcsPath, so we don't need to use URI to store the path.

I think GcsPath only take legal names and valid chars for gcs, and don't need to escape them.
If other contains any chars that is illegal in gcs, I think the current behavior (throwing) is right. Otherwise, the pipeline might read different files or write to different places.